plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/TypeJudgment.java

```diff
@@ -607,12 +612,55 @@ public TypeRef caseIdentifierRef(IdentifierRef idref) {
 							}
 
 							TypeUtils.sanitizeRawTypeRef(instanceofType);
-							intersectionTypes.add(instanceofType);
+							if (ioGuard.asserts == GuardAssertion.AlwaysHolds) {
+								allTypes.add(instanceofType);
+							} else if (ioGuard.asserts == GuardAssertion.NeverHolds) {
+								excludedTypes.add(instanceofType);
+							}
 						}
 					}
 				}
-				if (!intersectionTypes.isEmpty()) {
-					T = tsh.createIntersectionType(G, intersectionTypes.toArray(new TypeRef[intersectionTypes.size()]));
+
+				if (excludedTypes.isEmpty()) {
+					remainingTypes.addAll(allTypes);
+				} else {
+					for (TypeRef origType : allTypes) {
+						TypeRef remainingType = null;
+						if (origType instanceof UnionTypeExpression) {
+							final UnionTypeExpression union = (UnionTypeExpression) origType;
+							final Collection<TypeRef> newUnionTypes = new LinkedList<>();
+							for (TypeRef unionType : union.getTypeRefs()) {
+								for (TypeRef neverHoldingType : excludedTypes) {
+									if (!ts.equaltypeSucceeded(G, unionType, neverHoldingType)) {
+										newUnionTypes.add(unionType);
+									}
+								}
+							}
+							if (newUnionTypes.size() < union.getTypeRefs().size()) {
+								remainingType = tsh.createUnionType(G,
+										newUnionTypes.toArray(new TypeRef[newUnionTypes.size()]));
+							}
+						} else {
+
+							boolean removeType = false;
+							for (TypeRef neverHoldingType : excludedTypes) {
+								if (ts.equaltypeSucceeded(G, origType, neverHoldingType)) {
```

Comment:
I think `ts.subtypeSucceeded()` might be enough here (and also in line 634 above). This would then support cases like:
```javascript
class A {}
class BSuper {}
class B extends BSuper {}

function foo(p: A|B) {
	if (p instanceof BSuper) {
		return;
	}
	// XPECT type of 'p' --> A
	p;
}
```
But I'm not sure if this is valid in all cases.

Reply:
nice!